검색 페이지 UI 구현 & 검색 페이지 API 연결

park-moen opened the pull request:
## 💡 Linked Issues
- close: #64 

## 📖 구현 내용
- [x] API 통신 테스트
- [x] 검색 페이지 레이아웃 UI 구현하기
  - [x] SearchBar UI 구현하기
  - [x] SearchTab UI 구현하기
  - [x] SearchedPoster UI 구현하기 
- [x] API 연결 

## 🖼 구현 이미지
https://user-images.githubusercontent.com/57402711/213270131-c9f6d1bb-fa60-4ce2-8297-6e45f4f3c5cb.mov

## 반영 브랜치
`feature/search-page-UI/#64` -> `dev`

## ✅ PR 포인트 & 궁금한 점
- UI가 같아서 공통 컴포넌트로 사용할 수 있다고 생각했는데 막상 공통화하니 다른 API 정보에 대한 Type을 지정하기 어렵네요. ㅠ🥹🥹🥹 
- 멘토님 TypeScript 공통화 하는 꿀팁이 있을까요ㅠ ㅠ ㅠ ㅠ ㅠ

park-moen committed:
SearchBar UI 구현

park-moen committed:
formData 최상위 컴포넌트로 옮기기

park-moen committed:
SearchTab UI 컴포넌트 구현

park-moen committed:
사용자 검색 & 피드 검색 컴포넌트 구현

park-moen committed:
검색 페이지 전체 마진 값 적용 & FeedItem 높이 반응형 UI로 적용

live-small commented on src/utils/constants.ts:
```diff
@@ -19,3 +19,9 @@ export const DESKTOP_SCREEN = '1280px';
+
+// 카테고리 기본 정보
+export const DEFAULT_CATEGORY_INFORMATION = {
+  name: '노트북',
+  id: '63bd045193836272216d31bc',
```
카테고리 기본정보 상수화하는 건 좋은데, name 정도만 상수화해두고, id는 카테고리 상태값에서 찾아오는 게 어떨까??

park-moen replied:
기본값으로 매핑하면 좋겠다는 생각에 이렇게 만들었어요.

chunwookJoo commented on src/components/Search/SearchBar.tsx:
```diff
@@ -0,0 +1,55 @@
+  FieldErrors,
+  UseFormHandleSubmit,
+} from 'react-hook-form';
+import { MdCancel as CancelIcon } from 'react-icons/md';
```
아 이렇게 이름 변경해서 쓰는게 더 좋겠군용 👍

choibyeol commented on src/components/Search/SearchTab.tsx:
```diff
@@ -0,0 +1,38 @@
+const SearchTab = () => {
+  const [searchType, setSearchType] = useRecoilState(SearchState);
+  const buttonStyle = 'w-full p-2 text-TEXT_SUB_GRAY font-semibold cursor-pointer';
+  const selectedButtonType = 'border-b-2 border-BUTTON_TEXT_DISABLED';
```
p3;
이거 검색 버튼 선택 됐을 때 밑줄이라고 해야하나? 모바일 화면일 때 border-b 가로 길이 조금 줄이면 더 예쁠 것 같아용

choibyeol commented on src/components/Search/SearchTab.tsx:
```diff
@@ -0,0 +1,38 @@
+        >
+          사용자 검색
+        </button>
+      </div>
```
p4; refactor;
이부분 컴포넌트화 할 수 있을 것 같네요~

choibyeol commented on src/components/Search/SearchBar.tsx:
```diff
@@ -0,0 +1,55 @@
+            type="button"
+            className="absolute top-1/2 right-5 transform -translate-y-1/2"
+          >
+            <CancelIcon className="text-xl" />
```
p1;
canceIcon 누르면 검색어 없어지도록 해야할 것 같은데 액션이 없네요~~

chunwookJoo commented on src/components/Search/SearchTab.tsx:
```diff
@@ -0,0 +1,38 @@
+      <div className="w-1/2 text-center">
+        <button
+          name="all"
+          className={`${buttonStyle}  ${searchType === 'all' && selectedButtonType}`}
```
여기 스페이스 두칸 띄워져있나용?

live-small commented on src/components/common/Input/index.tsx:
```diff
@@ -0,0 +1,34 @@
+
+type InputPropsType = RegisterInputProps<SearchFormData>;
+
+const Input = ({
+  type,
+  placeholder,
+  register,
+  registerName,
+  registerRules,
+  style,
+  errors,
+  ...props
+}: InputPropsType) => {
+  return (
+    <div className={style?.container}>
+      <input
+        className={style?.input}
+        placeholder={placeholder}
+        type={type}
+        {...(register && register(registerName, registerRules))}
+        {...props}
+      />
+      {errors && <ErrorMessage errors={errors} />}
+    </div>
+  );
+};
+
```
🔥 리팩토링 

p4; 
이거 내가 만든 RegisterInput이랑 거의 똑같은데,  props 타입 달라서 못쓴거지?? 이거 확장해서 재사용하면 좋겠어

park-moen replied:
실패 ㅠ

choibyeol commented on src/components/common/Feed/index.tsx:
```diff
@@ -0,0 +1,48 @@
+}: FeedPropsType) => {
+  return (
+    <Link
+      className="border-HOVER cursor-pointer relative overflow-hidden w-1/3 h-10/12"
```
p3;
검색창과도 간격을 주고 피드간에도 간격을 조금 주는 게 예쁠 것 같아요~
p-1 mt-1 추가?

choibyeol commented on src/pages/Search/index.tsx:
```diff
@@ -1,5 +1,79 @@
+
+import { ExtractPostDataType, ExtractUserDataType } from '@/types/search';
+
+import SearchBar from '@/components/Search/SearchBar';
+import SearchTab from '@/components/Search/SearchTab';
+import SearchPostFeed from '@/components/Search/SearchPostFeed';
+
+import { SearchState } from '@/store/recoilSearchState';
+
+import { getSearchByType } from '@/Api/search';
+import SearchUserFeed from '@/components/Search/SearchUserFeed';
```
p4;
```ts
import { SearchState } from '@/store/recoilSearchState';

import { getSearchByType } from '@/Api/search';

import SearchBar from '@/components/Search/SearchBar';
import SearchTab from '@/components/Search/SearchTab';
import SearchPostFeed from '@/components/Search/SearchPostFeed';
import SearchUserFeed from '@/components/Search/SearchUserFeed';
```
순서로 수정해 주세요~

choibyeol commented on src/components/Search/SearchTab.tsx:
```diff
@@ -0,0 +1,38 @@
+    <section className="flex pt-4">
+      <div className="w-1/2 text-center">
+        <button
+          name="all"
```
p4;
피드 검색과 사용자 검색으로 나눠져 있으니까 all 보다는 feed 또는 post가 어울리지 않을까요?
수정하게 되면 recoil SearchState도 수정이 필요하겠네요

choibyeol commented on src/types/index.d.ts:
```diff
@@ -46,6 +46,30 @@ export type Post = {
   createdAt: string;
   updatedAt: string;
 };
+
+export type User = {
+  image: Optional<string>;
+  likes: Like[];
+  comments: string[];
+  createdAt: string;
+  banned: boolean;
+  email: string;
+  emailVerified: string;
+  _id: string;
+  fullName: string;
+  imagePublicId: string;
+  role: string;
+  updatedAt: string;
+  username: string;
+  _v: number;
+  posts: string[];
+  notifications: string[];
+  message: string[];
+  followers: string[];
+  following: string[];
+  isOnline: boolean;
+};
```
user type 한 3명이 정의한 것 같은데...😂
필수 아닌 속성들도 있어서 타입 통일이 필요할 것 같네요

chunwookJoo reviewed (approved):
수고용

choibyeol reviewed (approved):
검색 구현 고생했어 👍👍
cancel 액션은 꼭 추가해 주세용~~
나머지 수정 사항은 바쁘면 일단 나중에 처리해도 될 것 같습니당

live-small commented on src/components/Search/SearchPostFeed.tsx:
```diff
@@ -0,0 +1,33 @@
+
+const SearchFeedSection = ({ searchedPost }: { searchedPost: ExtractPostDataType[] }) => {
+  const categoryNameAndIdList = useRecoilValue(categoryState);
+  const AddPropsFeedComponent = searchedPost.map((aSearchedPost) => {
```
p2; 
함수같이 보입니당 ! 변수명 수정부탁드려용

live-small commented on src/components/Search/SearchTab.tsx:
```diff
@@ -0,0 +1,38 @@
+  const buttonStyle = 'w-full p-2 text-TEXT_SUB_GRAY font-semibold cursor-pointer';
+  const selectedButtonType = 'border-b-2 border-BUTTON_TEXT_DISABLED';
+
+  const onClickTab = ({ currentTarget: { name } }: React.MouseEvent<HTMLButtonElement>) =>
```
오잉 여기 뒤에 스페이스 뭔가용? 
```suggestion
  const onClickTab = ({ currentTarget: { name }}: React.MouseEvent<HTMLButtonElement>) =>
```

park-moen replied:
동일한 스페이스로 간격을 인지하고 자동완성 되네요!

live-small commented on src/components/Search/SearchUserFeed.tsx:
```diff
@@ -0,0 +1,28 @@
+import { MY_PAGE } from '@/utils/constants';
+
+const SearchUserFeed = ({ searchedUser }: { searchedUser: ExtractUserDataType[] }) => {
+  const AddPropsFeedComponent = searchedUser.map((aSearchedUser) => {
```
p2; 
add로 시작하니까 함수같습니다!!! 
컴포넌트 반환하니까 앞에 동사는 지양해주세용

live-small commented on src/pages/Search/index.tsx:
```diff
@@ -1,5 +1,79 @@
+
+        setSearchedPost([...extractPostInformation]);
+        setSearchedUser([]);
+      } else {
```
else말고 early return 해주세여

park-moen committed:
(코드 리뷰 적용)검색 페이지 UI 구현 & 검색 페이지 API 연결

park-moen committed:
검색 페이지 UI 구현 & 검색 페이지 API 연결 이슈 충돌 해결

; Conflicts:
;	src/pages/Search/index.tsx
;	src/types/index.d.ts
;	src/utils/constants.ts

park-moen committed:
검색 페이지 UI 구현 & 검색 페이지 API 연결 충돌 해결 후 통합

park-moen committed:
불필요한 콘솔 모두 제거
